blackbird_python/blackbird/listener.py

```diff
@@ -447,26 +445,39 @@ def exitStatement(self, ctx: blackbirdParser.StatementContext):
             self._program._operations.append(operation)
 
     def enterForloop(self, ctx: blackbirdParser.ForloopContext):
-        self._infor = True
+        self._in_for = True
 
     def exitForloop(self, ctx: blackbirdParser.ForloopContext):
-        self._infor = False
+        self._in_for = False
 
-        if ctx.INT():
-            _FORVAR = range(int(ctx.INT().getText()))
+        if ctx.rangeval():
+            for_var = range(*[
+                int(c.getText()) for c in ctx.rangeval().getChildren()
+                if c.getText() != ":"
+            ])
         elif ctx.vallist():
-            _FORVAR = ctx.vallist().getText().split(',')
+            for_var = [
+                _expression(c.expression()) for c in ctx.vallist().getChildren()
+                if isinstance(c, blackbirdParser.ValContext)
+            ]
 
-        for var in _FORVAR:
+        for var in for_var:
             if ctx.NAME():
                 try:
-                    var = PYTHON_TYPES[ctx.vartype().getText()](var)
+                    new_var = PYTHON_TYPES[ctx.vartype().getText()](var)
+                    if new_var != var:
+                        raise ValueError
                 except ValueError:
-                    raise ValueError(f"invalid value {var}; must be {ctx.vartype().getText()}")
-                _VAR[ctx.NAME().getText()] = var
+                    raise ValueError("invalid value {}; must be {}".format(var, ctx.vartype().getText()))
+
+                _VAR[ctx.NAME().getText()] = new_var
+
             for statement in ctx.statement_list:
                 self.exitStatement(statement)
-        _FORVAR = []
+
+        self._program._forvar[ctx.NAME().getText()] = np.array(for_var)
+        if ctx.NAME():
+            del _VAR[ctx.NAME().getText()]
```

Comment:
This is to delete the temp for loop variable?

Reply:
Yes, otherwise the last element of the for-loop list will be left as a "free" variable, which isn't a huge problem but it's nicer if it's confined to the for-loop. It could cause issues later.